disdat/db_target.py

```diff
@@ -0,0 +1,611 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from luigi.target import Target
+import pyodbc
+import logging
+import contextlib
+import pandas as pd
+
+_logger = logging.getLogger(__name__)
+
+#
+# Your database may not support schema creation to logically separate
+# tables in different data contexts.   This should be a configuration option.
+#
+
+ENABLE_DISDAT_SCHEMAS = False
+UUID_PREFIX = 8 # number of chars to use as uuid prefix in phys_name
+
+
+@contextlib.contextmanager
+def get_connection(dsn):
+    conn = pyodbc.connect(dsn=dsn)
+    yield conn
+    conn.close()
+
+
+@contextlib.contextmanager
+def get_cursor(dsn):
+    conn = pyodbc.connect(dsn=dsn)
+    cursor = conn.cursor()
+    yield cursor
+    cursor.commit()
+    cursor.close()
+    conn.close()
+
+
+def single_query(dsn, query, cursor=None):
+    """
+
+    Args:
+        dsn (unicode) : The database configuration dsn
+        query (unicode) : Raw query text
+        cursor:
+
+    Returns:
+        df (pd.DataFrame) or desc (str)
+
+    """
+
+    #print ("query: Running query:{}".format(query))
+
+    get_rows = True
+    if "DROP" in query or "CREATE" in query:
+        get_rows = False
+
+    def _query(cursor, query):
+        cursor.execute(query)
+        desc = cursor.description
+        df = None
+        if get_rows:
+            rows = cursor.fetchall()
+            col_names = [x[0] for x in cursor.description]
+            df = pd.DataFrame.from_records(rows, columns=col_names)
+        return desc, df
+
+    if cursor is None:
+        with get_cursor(dsn) as cursor:
+            desc, df = _query(cursor, query)
+    else:
+        desc, df = _query(cursor, query)
+
+    if get_rows:
+        return df
+    else:
+        return desc
+
+
+def multi_tx_query(dsn, queries):
+    """
+    Issue multiple queries in a single transaction.  We depend
+    on our get_cursor() function to call cursor.commit().
+
+    Args:
+        dsn (unicode) : The database configuration dsn
+        queries (list:unicode) : Query texts
+
+    Returns:
+        list:df (pd.DataFrame) or list:desc (str)
+
+    """
+    results = []
+    with get_cursor(dsn) as cursor:
+        for q in queries:
+            results.append(single_query(dsn, q, cursor))
+
+    return results
+
+
+def table_exists_vertica(dsn, schema, table):
+    """
+    Determine whether a table exists in a Vertica DataBase
```

Comment:
does it only apply to Vertica?

Reply:
Yes.   Later, one possibility is to make the DBTarget be parameterized by SQL statements for other DB's.